core/src/main/java/org/pentaho/di/core/row/value/ValueMetaTimestamp.java

```diff
@@ -0,0 +1,709 @@
+          default:
+            throw new KettleValueException( toString() + " : Unknown storage type " + storageType + " specified." );
+        }
+      case TYPE_BOOLEAN:
+        throw new KettleValueException( toString() + " : I don't know how to convert a boolean to a timestamp." );
+      case TYPE_BINARY:
+        throw new KettleValueException( toString() + " : I don't know how to convert a binary value to timestamp." );
+      case TYPE_SERIALIZABLE:
+        throw new KettleValueException( toString()
+          + " : I don't know how to convert a serializable value to timestamp." );
+
+      default:
+        throw new KettleValueException( toString() + " : Unknown type " + type + " specified." );
+    }
+  }
+
+  public int compare( Object data1, Object data2 ) throws KettleValueException {
+    Timestamp timestamp1 = getTimestamp( data1 );
+    Timestamp timestamp2 = getTimestamp( data2 );
+    int cmp = 0;
+    if ( timestamp1 == null ) {
+      if ( timestamp2 == null ) {
+        cmp = 0;
+      } else {
+        cmp = -1;
+      }
+    } else if ( timestamp2 == null ) {
+      cmp = 1;
+    } else {
+      cmp = timestamp1.compareTo( timestamp2 );
+    }
+    if ( isSortedDescending() ) {
+      return -cmp;
+    } else {
+      return cmp;
+    }
+  }
+
+  protected Timestamp convertBigNumberToTimestamp( BigDecimal bd ) {
+    if ( bd == null ) {
+      return null;
+    }
+    return convertIntegerToTimestamp( bd.longValue() );
+  }
+
+  protected Timestamp convertNumberToTimestamp( Double d ) {
+    if ( d == null ) {
+      return null;
+    }
+
+    return convertIntegerToTimestamp( d.longValue() );
+  }
+
+  protected Timestamp convertIntegerToTimestamp( Long longValue ) {
+    if ( longValue == null ) {
+      return null;
+    }
+
+    Long nanos = longValue;
+
+    if ( Const.KETTLE_TIMESTAMP_NUMBER_CONVERSION_MODE_MILLISECONDS.equalsIgnoreCase( conversionMode ) ) {
+      // Convert milliseconds to nanoseconds!
+      nanos *= 1000000L;
+    }
+
+    long ss = TimeUnit.SECONDS.convert( nanos, TimeUnit.NANOSECONDS );
+    long ms = TimeUnit.MILLISECONDS.convert( nanos, TimeUnit.NANOSECONDS );
+    long ns = TimeUnit.NANOSECONDS.convert( ss, TimeUnit.SECONDS );
+    int leftNs = (int) ( nanos - ns );
+
+    Timestamp timestamp = new Timestamp( ms );
+    timestamp.setNanos( leftNs );
+    return timestamp;
+  }
+
+  protected synchronized Timestamp convertStringToTimestamp( String string ) throws KettleValueException {
+    // See if trimming needs to be performed before conversion
+    //
+    string = Const.trimToType( string, getTrimType() );
+
+    if ( Utils.isEmpty( string ) ) {
+      return null;
+    }
+    Timestamp returnValue;
+    try {
+      returnValue = Timestamp.valueOf( string );
+    } catch ( IllegalArgumentException e ) {
+      try {
+        returnValue = (Timestamp) getDateFormat().parse( string );
+      } catch ( ParseException ex ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)